src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx

```diff
@@ -122,312 +62,536 @@ const SicpNavigationBar: React.FC = () => {
+        setIndexAutoCompleteCouldShow(false);
+      }
+    };
+    document.addEventListener('click', handleClickOutside);
+    return () => {
+      document.removeEventListener('click', handleClickOutside);
+    };
+  }, [autocompleteMenuRef, searchInputRef, indexSearchInputRef]);
+  const menu = (
+    searchInput: string,
+    isSubMenueHidden: (searchInput: string) => Boolean,
+    getResults: () => any[],
+    buildMenuEntry: (searchResult: any, index: number) => React.ReactNode,
+    buildMenuWith: (
+      children: React.ReactNode,
+      ref: React.RefObject<HTMLDivElement>
+    ) => React.ReactNode
+  ) => {
+    if (isSubMenueHidden(searchInput)) {
+      return <></>;
+    }
+    const children = getResults().map(buildMenuEntry);
+    return buildMenuWith(children, autocompleteMenuRef);
   };
 
-  const handleIndexSearchButton = (str: string) => {
-    handleOpenSearch();
-    setDisplayedQuery(str);
-    const SearchUrl = '..';
-    const tem = [];
-    const ans = queryTrie(indexTrie, str.toLowerCase());
-    if (ans == null) {
-      tem.push({ title: 'no result found', url: '' });
-    } else {
-      const pure = ans['pureIndex'];
-      for (let i = 0; i < pure.length; i++) {
-        // pure[i][0] is sth like /sicpjs/3.3.3; slice out the /sicpjs/
-        tem.push({
-          title: pure[i][0].slice(8) + ': ' + ans['value'],
-          url: SearchUrl + pure[i][0] + pure[i][1]
-        });
+  const userSearchResultSubMenu = (searchInput: string) => {
+    const focusResult = (result: string) => {
+      result = result.replaceAll('\n', ' ').toLowerCase();
+      const startIndex = result.indexOf(searchInput);
+      let start = startIndex;
+      while (start > 0) {
+        if (result[start - 1].match(/[^a-zA-Z, _]/)) {
+          console.log('break at ' + (start - 1) + ' ' + result[start - 1]);
+          break;
+        }
+        start--;
       }
-      const subindex = ans['subIndex'];
-      for (let i = 0; i < subindex.length; i++) {
-        tem.push({
-          title: subindex[i]['id'][0].slice(8) + ': ' + ans['value'] + ': ' + subindex[i]['value'],
-          url: SearchUrl + subindex[i]['id'][0] + subindex[i]['id'][1]
-        });
+      const endIndex = startIndex + searchInput.length;
+      let end = endIndex;
+      while (end < result.length) {
+        if (result[end].match(/[^a-zA-Z _,]/)) {
+          console.log('break at ' + end + ' ' + result[end]);
+          break;
+        }
+        end++;
       }
-    }
-
-    setQueryResult(tem);
-  };
+      let subStr = result.slice(start, end);
+      console.log('subStr is ' + subStr);
+      if (start > 0) {
+        subStr = '...' + subStr;
+      }
+      if (end < result.length) {
+        subStr = subStr + '...';
+      }
+      return subStr;
+    };
+
+    const isUserSearchSubMenuHidden = (searchInput: string) => {
+      return !(isSubmenuVisible === searchInput);
+    };
+
+    const getUserSearchResults = () => {
```

Comment:
Is there the need for this additional function that basically just returns `results` (possible to use `results` directly instead?)

Reply:
This is for the template method "menu" to works, which need to take in 4 functions